File: frontend/src/types/task.ts
/**
 * TypeScript Type Definitions
 * Strictly aligned with Backend Pydantic Models & API Schemas
 */

export type TaskStatus =
  | 'pending'
  | 'running'
  | 'paused_for_approval'
  | 'completed'
  | 'failed'
  | 'cancelled';

export type WorkflowPhase =
  | 'initialized'
  | 'repository_analysis'
  | 'planning'
  | 'coding'
  | 'testing'
  | 'debugging'
  | 'review'
  | 'commit'
  | 'pull_request'
  | 'finished';

export type ApprovalStatus =
  | 'not_required'
  | 'pending'
  | 'approved'
  | 'rejected';

export type ToolCategory =
  | 'repository'
  | 'editing'
  | 'execution'
  | 'git'
  | 'github';

export interface PlanStep {
  step_id: number;
  title: string;
  description: string;
  target_files: string[];
  status: 'pending' | 'in_progress' | 'completed' | 'failed' | string;
  completed: boolean;
  notes?: string | null;
}

export interface AgentPlan {
  objective: string;
  architecture_overview: string;
  steps: PlanStep[];
  current_step_index: number;
}

export interface ToolExecutionRecord {
  call_id: string;
  tool_name: string;
  input_args: Record<string, any>;
  output?: string | null;
  error?: string | null;
  exit_code?: number | null;
  execution_time_ms: number;
  requires_approval: boolean;
  approval_status: ApprovalStatus;
  timestamp: string;
}

export interface TestExecutionSummary {
  command: string;
  passed: boolean;
  total_tests: number;
  failures: number;
  errors: number;
  stdout: string;
  stderr: string;
  exit_code: number;
  duration_seconds: number;
}

export interface PendingApproval {
  approval_id?: string | null;
  id?: string | null;
  action_type: string;
  tool_name: string;
  action_payload: Record<string, any>;
  payload: Record<string, any>;
  reason: string;
  status: ApprovalStatus;
  reviewer_feedback?: string | null;
  created_at: string;
  requested_at?: string | null;
}

export interface CreateTaskRequest {
  repository_url: string;
  user_instruction: string;
  base_branch?: string;
  working_branch?: string;
  max_retries?: number;
}

export interface TaskResponse {
  id: string;
  repository_url: string;
  user_instruction: string;
  status: TaskStatus;
  current_phase: WorkflowPhase;
  base_branch: string;
  working_branch: string;
  commit_sha?: string | null;
  pull_request_url?: string | null;
  plan?: AgentPlan | null;
  retry_count: number;
  created_at: string;
  updated_at: string;
  error_message?: string | null;
}

export interface TaskDetailResponse extends TaskResponse {
  tool_history: ToolExecutionRecord[];
  modified_files: string[];
  test_results: TestExecutionSummary[];
  pending_approval?: PendingApproval | null;
}

export interface ApprovalDecisionRequest {
  approved: boolean;
  feedback?: string | null;
}

export interface TaskEvent {
  task_id: string;
  event_type: string;
  phase: WorkflowPhase;
  message: string;
  data: Record<string, any>;
  timestamp: string;
}

export interface TaskDiffResponse {
  task_id: string;
  diff: string;
}

File: app/models/task.py
"""API Request, Response, and Event Models."""

from datetime import datetime, timezone
from typing import Any, Dict, List, Optional
from pydantic import BaseModel, Field, HttpUrl

from app.models.enums import ApprovalStatus, TaskStatus, WorkflowPhase
from app.models.state import (
    AgentPlan,
    PendingApproval,
    TestExecutionSummary,
    ToolExecutionRecord,
)


def utc_now() -> datetime:
    """Return timezone-aware current UTC datetime."""
    return datetime.now(timezone.utc)


class CreateTaskRequest(BaseModel):
    """Payload to launch a new software engineering task."""
    repository_url: str = Field(
        ...,
        description="Public or authenticated Git/GitHub repository URL",
        examples=["https://github.com/fastapi/fastapi"],
    )
    user_instruction: str = Field(
        ...,
        description="Natural language instruction for the agent",
        min_length=5,
        examples=["Add JWT authentication endpoints and write unit tests."],
    )
    base_branch: str = Field(default="main", description="Target base branch")
    working_branch: Optional[str] = Field(
        default=None,
        description="Custom branch name (auto-generated if omitted)",
    )
    max_retries: int = Field(
        default=5,
        ge=1,
        le=15,
        description="Maximum test & debug retry loops before asking human input",
    )


class TaskResponse(BaseModel):
    """Summary of an engineering task."""
    id: str
    repository_url: str
    user_instruction: str
    status: TaskStatus
    current_phase: WorkflowPhase
    base_branch: str
    working_branch: str
    commit_sha: Optional[str] = None
    pull_request_url: Optional[str] = None
    plan: Optional[AgentPlan] = None
    retry_count: int = 0
    created_at: datetime
    updated_at: datetime
    error_message: Optional[str] = None


class TaskDetailResponse(TaskResponse):
    """Comprehensive task details including full execution history."""
    tool_history: List[ToolExecutionRecord] = Field(default_factory=list)
    modified_files: List[str] = Field(default_factory=list)
    test_results: List[TestExecutionSummary] = Field(default_factory=list)
    pending_approval: Optional[PendingApproval] = None


class ApprovalDecisionRequest(BaseModel):
    """Payload to approve or reject a paused action."""
    approved: bool
    feedback: Optional[str] = Field(
        default=None,
        description="Feedback or corrective guidance if rejecting or adjusting",
    )


class TaskEvent(BaseModel):
    """Event emitted over SSE/WebSocket streams."""
    task_id: str
    event_type: str
    phase: WorkflowPhase
    message: str
    data: Dict[str, Any] = Field(default_factory=dict)
    timestamp: datetime = Field(default_factory=utc_now)
